py/desitransfer/daemon.py

```diff
@@ -273,108 +232,92 @@ def exposure(self, d, link, status):
+            log.debug("os.chmod('%s', 0o%o)", destination_night, dir_perm)
+            if not self.test:
+                os.makedirs(destination_night, exist_ok=True)
+                os.chmod(destination_night, dir_perm)
+        #
         # Has exposure already been transferred?
         #
         if not os.path.isdir(staging_exposure) and not os.path.isdir(destination_exposure):
             cmd = rsync(os.path.join(d.source, night, exposure), staging_exposure)
+            log.debug(' '.join(cmd))
             if self.test:
-                log.debug(' '.join(cmd))
                 rsync_status = '0'
             else:
                 rsync_status, out, err = _popen(cmd)
         else:
             log.debug('%s already transferred.', staging_exposure)
-            rsync_status = 'done'
+            return
         #
         # Transfer complete.
         #
         if rsync_status == '0':
             log.debug("status.update('%s', '%s', 'rsync')", night, exposure)
-            status.update(night, exposure, 'rsync')
-            #
-            # Check permissions.
-            #
-            lock_directory(staging_exposure, self.test)
-            #
-            # Verify checksums.
-            #
+            if not self.test:
+                status.update(night, exposure, 'rsync')
+        else:
+            log.critical('rsync problem (status = %s) detected for %s/%s, check logs!',
+                         rsync_status, night, exposure)
+            log.error('rsync STDOUT = %s', out)
+            log.error('rsync STDERR = %s', err)
+            log.debug("status.update('%s', '%s', 'rsync', failure=True)", night, exposure)
+            status.update(night, exposure, 'rsync', failure=True)
+        #
+        # Check permissions.
+        #
+        log.debug("lock_directory('%s', %s)", staging_exposure, str(self.test))
+        lock_directory(staging_exposure, self.test)
+        #
+        # Verify checksums.
+        #
+        checksum_file = os.path.join(staging_exposure,
+                                     d.checksum.format(night=night, exposure=exposure))
+        self.checksum(checksum_file, status)
+        #
+        # Move data into DESI_SPECTRO_DATA.
+        #
+        if not os.path.isdir(destination_exposure):
+            log.debug("shutil.move('%s', '%s')", staging_exposure, destination_night)
+            if not self.test:
+                shutil.move(staging_exposure, destination_night)
+
+    def checksum(self, checksum_file, status):
+        """Verify checksum associated with `exposure` and report status.
 
-            checksum_file = os.path.join(staging_exposure,
-                                         d.checksum.format(night=night, exposure=exposure))
+        Parameters
+        ----------
+        checksum_file : :class:`str`
+            The checksum file.
+        status : :class:`desitransfer.status.TransferStatus`
+            The associated status object.
```

Comment:
When a function modifies one of its inputs, it's nice to comment on that in the docstring.  In this case, it could clarify that this function modified `status` rather than returning a checksum succes/error/status.

Reply:
I'm not sure what you mean by "modified" in this case.  The method *uses* the status object, but it doesn't really *modify* it.

Reply:
I guess you just mean that the result of the check is reported via log messages and the status object, rather than via return value?

Reply:
Perhaps I misunderstood.  I thought the `status` object was tracking state that was updated via `status.update(night, exposure, ...)` when this function was called.  If the only effect of calling this function is printing log messages, with no changes to the `status` object itself and no return value, then this docstring is fine.

Reply:
OK, then we definitely don't agree on the definition of "modify".  The status object does not have *state*, in the object-oriented sense, that can be modified.  It is a fancy way to *write* the state to a file.  That's why I say the status object is used but not modified.